src/Ocean/HydrostaticBoussinesq/HydrostaticBoussinesqModel.jl

```diff
@@ -635,13 +629,20 @@ function update_auxiliary_state_gradient!(
     indefinite_stack_integral!(dg, m, Q, A, t, elems) # bottom -> top
     reverse_indefinite_stack_integral!(dg, m, Q, A, t, elems) # top -> bottom
 
-    # project w(z=0) down the stack
-    # [1] to convert from range to integer
-    # copy_stack_field_down! doesn't like ranges
-    # eventually replace this with a reshape and broadcast
+    Nq, Nqk, _, _, nelemv, _, nelemh, _ = basic_grid_info(dg)
+
+    # We are unable to use vars (ie A.w) for this because this operation will
+    # return a SubArray, and adapt (used for broadcasting along reshaped arrays)
+    # has a limited recursion depth for the types allowed.
     index_w = varsindex(vars_state_auxiliary(m, FT), :w)[1]
     index_wz0 = varsindex(vars_state_auxiliary(m, FT), :wz0)[1]
-    copy_stack_field_down!(dg, m, A, index_w, index_wz0, elems)
+
+    # project w(z=0) down the stack
+    boxy_w = reshape(A.data[:,index_w,:], Nq^2, Nqk, 1, nelemv, nelemh)
+    flat_w = @view boxy_w[:, end, :, end, :]
+    flat_wz0 = reshape(flat_w, Nq^2, 1, 1, 1, nelemh)
+    boxy_wz0 = reshape(A.data[:,index_wz0,:], Nq^2, Nqk, 1, nelemv, nelemh)
```

Comment:
I don't think this will work, you are creating an array slice not a view, so you are not modifying the original array but a newly created array.

Reply:
Also, you haven't accounted for the ghost elements, so I don't think it will work in parallel.

Reply:
Possibly it would work in parrallel once you fix the views with:
```julia
    Nq, Nqk, _, _, nelemv, nelemh, _, _ = basic_grid_info(dg)
```
above, though you will be copying down the ghost elements too.

Reply:
ah, I think you are right. This worked on my GPU, but we need to think of a more general solution.

Reply:
I think you need to reshape, then do the views (like I did here: https://github.com/CliMA/ClimateMachine.jl/pull/1301#issuecomment-651313057)

The reshape returns a new CuArray (pointing to the same data) which is why you miss the double indirection problem.

Reply:
> The reshape returns a new CuArray (pointing to the same data) which is why you miss the double indirection problem.

At least I believe that's what's happening... don't have a gpu box to test on at this moment.

Reply:
> I think you need to reshape, then do the views (like I did here: [#1301 (comment)](https://github.com/CliMA/ClimateMachine.jl/pull/1301#issuecomment-651313057))
> 
> The reshape returns a new CuArray (pointing to the same data) which is why you miss the double indirection problem.

We couldn't quite figure out how to phrase the problem in that way, but I think you are right.

Reply:
What about:
```suggestion
data = reshape(A.data, Nq^2, Nqk, num_aux_state, 1, nelemv, nelemh)
flat_wz0 = @view data[index_w, :, end:end, :, end:end, :]
boxy_wz0 = @view data[index_wz0, :, :, :, :, :]
```

Reply:
This (original code) doesn't work, not even on a CPU. so we need to make some kind of further change for sure

@jkozdon can you refresh me on the `MPIStateArray.data` layout? I thought it was `["GL points in element", "states", "elements"]` but that doesn't match up with your `views` (or even the first `reshape`, shouldn't the `1` be a `:`?).

Reply:
Looks good to me lurking  - is there an extra ```,1,``` after ```num_aux_state``` or is that needed for some more magic?

Reply:
> Looks good to me lurking - is there an extra `,1,` after `num_aux_state` or is that needed for some more magic?

Ah it is an extra `,1,`! that's what my parenthetical was about

Reply:
o.o I think @jkozdon is editting it, the indices just jumped

Reply:
> This (original code) doesn't work, not even on a CPU. so we need to make some kind of further change for sure

Couple typos for sure. Updated my suggestion.

> @jkozdon can you refresh me on the `MPIStateArray.data` layout? I thought it was `["GL points in element", "states", "elements"]` but that doesn't match up with your `views` (or even the first `reshape`, shouldn't the `1` be a `:`?).

data ordering is `[dof, state, elements]`

Reply:
just pushed a commit based on that fix, worked on the CPU, about to try it on the GPU

Reply:
hmm didn't succeed on GPU due to the reshape dimensions not matching the `array.data`: https://gist.github.com/climabot/4a0358efe092e8b8015870c489a0e2df#file-out_10192381

any ideas @jkozdon? it works fine on CPU